### PR TITLE
C++17 (C++1z) enablement in ROOT, LLVM, Interpreter

### PR DESCRIPTION
**Pull request by davidlt**
This patch set enabled C++17 (C++1z) support in ROOT. This will require at least GCC 7.1.0 (soon to be released, probably in a month). GCC 7 fully implements C++17 language features.

CMSSW builds using GCC 6.2.0, GCC 6.3.0 and GCC 7.0.1 are all compiled with C++17.

Note that biggest issues came from `TString` and `std::string` interaction due to `std::string_view` conversion operator in `TString`. This makes mixing `TString` and `std::string` painful. Thus to avoid ambiguity errors `std::string_view` operator is now marked as explicit. Otherwise in a lot of places I had to use `static_cast<>` to pick one of conversion operators (which makes the patch bigger).

### CODE REVIEW

**Comment by davidlt** [comment]
Build tested using CMSSW `root.spec` on ppc64le (POWER8), CentOS 7.3, GCC 7.0.1 (a week old build).

**Comment by vgvassilev** [comment]
@davidlt, could you fix the clang-format issues?

File: core/dictgen/src/BaseSelectionRule.cxx [hunk 1]
```diff
@@ -317,7 +317,7 @@ BaseSelectionRule::EMatchType BaseSelectionRule::Match(const clang::NamedDecl *d
          if (!patternMatched &&
                D &&
                //ROOT::TMetaUtils::IsStdDropDefaultClass(*D)) {
-               0 != TClassEdit::IsSTLCont(name)) {
+               0 != TClassEdit::IsSTLCont(std::string_view(name.data(), name.size()))) {
```
Comment (Axel-Naumann) [inline core/dictgen/src/BaseSelectionRule.cxx]: Why do we need to construct a temporary `string_view` here, explicitly, from a `std::string`?

File: core/foundation/inc/RStringView.h [hunk 2]
```diff
@@ -29,7 +29,11 @@
 namespace std {
 
    template<class _CharT, class _Traits = std::char_traits<_CharT> >
+// What happens if we don't have std::basic_string_view and
+// std::experimental::basic_string_view?
```
Comment (Axel-Naumann) [inline core/foundation/inc/RStringView.h]: Then ROOT injects its own `string_view` into `std::`. Could you remove the comment?

File: core/foundation/inc/RStringView.h [hunk 3]
```diff
@@ -47,7 +51,9 @@ namespace std {
 #ifndef R__HAS_STOD_STRING_VIEW
    inline double stod(std::string_view str, size_t *pos)
    {
-      return std::stod(str.to_string(),pos);
+// to_string() method didn't make into C++17, but the following should be compatible
+// with std::sting_view and std::experimental::string_view.
```
Comment (Axel-Naumann) [inline core/foundation/inc/RStringView.h]: That's a good comment for the change, but a bit out of context for the code. Just remove it, please.

**Comment by davidlt** [comment]
Could you provide instructions (or point to page) to clang-format check scripts, or how ROOT runs clang-format?

**Comment by vgvassilev** [comment]
@davidlt, it is in the log.

**Comment by pcanal** [comment]
When building with gcc7, I am getting:
```
In file included from /local2/scratch/pcanal/build/bin/../lib/gcc/x86_64-pc-linux-gnu/7.0.1/../../../../include/c++/7.0.1/bits/node_handle.h:39:
/local2/scratch/pcanal/build/bin/../lib/gcc/x86_64-pc-linux-gnu/7.0.1/../../../../include/c++/7.0.1/optional:1032:27: error: use of class template 'optional' requires template arguments
  template <typename _Tp> optional(_Tp) -> optional<_Tp>;
                          ^
/local2/scratch/pcanal/build/bin/../lib/gcc/x86_64-pc-linux-gnu/7.0.1/../../../../include/c++/7.0.1/optional:451:11: note: template is declared here
    class optional
          ^
/local2/scratch/pcanal/build/bin/../lib/gcc/x86_64-pc-linux-gnu/7.0.1/../../../../include/c++/7.0.1/optional:1032:40: error: expected ';' at end of declaration
  template <typename _Tp> optional(_Tp) -> optional<_Tp>;
                                       ^
/local2/scratch/pcanal/build/bin/../lib/gcc/x86_64-pc-linux-gnu/7.0.1/../../../../include/c++/7.0.1/optional:1032:41: error: cannot use arrow operator on a type
  template <typename _Tp> optional(_Tp) -> optional<_Tp>;
                                        ^
Error: ../rootcling_stage1/src/rootcling_stage1: Linkdef compilation failure
```

**Comment by davidlt** [comment]
I completely forgot about it. I have this change reverted in CMSSW toolchain for now. This change was done in January: https://patchwork.ozlabs.org/patch/721403/

https://cplusplus.github.io/LWG/lwg-active.html#2825

File: core/foundation/src/TClassEdit.cxx [hunk 4]
```diff
@@ -1342,7 +1342,7 @@ bool TClassEdit::IsStdClass(const char *classname)
 bool TClassEdit::IsVectorBool(const char *name) {
    TSplitType splitname( name );
 
-   return ( TClassEdit::STLKind( splitname.fElements[0] ) == ROOT::kSTLvector)
+   return ( TClassEdit::STLKind( std::string_view(splitname.fElements[0].data(), splitname.fElements[0].size()) ) == ROOT::kSTLvector)
```
Comment (Axel-Naumann) [inline core/foundation/src/TClassEdit.cxx]: Given how often we need to handle the `STLKind` constructor, maybe you could add an overload taking a `string_view` (if the disambiguation is really needed)?

File: core/foundation/src/TClassEdit.cxx [hunk 5]
```diff
@@ -1871,8 +1871,10 @@ class NameCleanerForIO {
       if (!mother) return false;
       bool isSTLContOrArray = true;
       while (nullptr != mother){
-         auto stlType = TClassEdit::IsSTLCont(mother->fName+"<>");
-         isSTLContOrArray &= ROOT::kNotSTL != stlType || TClassEdit::IsStdArray(mother->fName+"<");
+         auto tmpStr = mother->fName+"<>";
+         auto stlType = TClassEdit::IsSTLCont(std::string_view(tmpStr.data(), tmpStr.size()));
```
Comment (Axel-Naumann) [inline core/foundation/src/TClassEdit.cxx]: I still don't understand why we need to provide explicit casts to `string_view` for `IsSTLCont`, `IsStdArray` etc. Isn't passing `tmpStr` as argument enough?

File: core/multiproc/inc/TMPWorkerExecutor.h [hunk 6]
```diff
@@ -110,7 +110,7 @@ class TMPWorkerExecutor : public TMPWorker {
          MPSend(s, MPCode::kFuncResult, fReducedResult);
       } else {
          reply += ": unknown code received: " + std::to_string(code);
-         MPSend(s, MPCode::kError, reply.data());
+         MPSend(s, MPCode::kError, reply.c_str());
```
Comment (Axel-Naumann) [inline core/multiproc/inc/TMPWorkerExecutor.h]: Thank you!

**Comment by Axel-Naumann** [comment]
@davidlt - we have 7 more days before we'll branch v6-10. It would be *really* nice if we could merge this before. Do you think you can help us with that?

**Comment by pcanal** [comment]
@axel I have actually been looking at this and I am investigating an alternative (for TString).

**Comment by Axel-Naumann** [comment]
@pcanal I thought that's whay @davidlt had tried? Hmm right, I guess `op string` wasn't explicit in his case. Anyway - I'd really like to have this in for 6.10: 6.10 should absolutely support GCC 7, ideally with C++17.

**Comment by davidlt** [comment]
`operator std::string` was not explicit in my case, but I don't think that alone would help. You still have ambiguity as you have `operator const char*` and `operator std::string_view`. Both of these are available as arguments to various `std::string` methods/ctors.

I am looking again at this patchset.  The changes to `core/foundation/src/TClassEdit.cxx` are no more needed. I am working on updating this PR.

Note, this still doesn't solve `optional` issue.

**Comment by pcanal** [comment]
so far in my build with the operator I mentioned plus an operator+=(string,TString), I need almost no modification to the ROOT source ...

**Comment by davidlt** [comment]
If you provide also a global `operator+=(string, TString)`, but not in all cases. E.g., passing `TString` to `std::string` ctor will continue to yield ambiguity errors. While it might lower number of lines we need to change in ROOT the users (who enable C++17) are still affected. Changes to ROOT are minimal. Where are just a few places where we need explicitly tell to compiler that we want `std::string_view`.

**Comment by pcanal** [comment]
With the set of operator I have tried, the change in ROOT are limited to the one call to stod.

> E.g., passing TString to std::string ctor will continue to yield ambiguity errors.

Actually this is one of the case that is solved by adding the explicit operator std::string.

The other case that are left are interfaces that takes *both* a const char* and a std::string_view ... we don't (seem) to have this case in the ROOT code ...

**Comment by pcanal** [comment]
Merged in with the change in operator I mentioned and a couple of tweaks.

Thanks.
Philippe.

**Comment by davidlt** [comment]
Hmm... You would also need `operator=(string,TString)`.

Does `explicit operator std::string` cover all cases?

I just checked issues with optional and they are caused because `rootcling_stage1` runs in C++14 mode, but no C++17. Currently I don't see where you set a language options for `rootcling_stage1`.

**Comment by pcanal** [comment]
> Hmm... You would also need operator=(string,TString).
Apparently not ....

The following code compiles
```
void tester() {
  TString str;
  std::string stdstr;
  stdstr = str;
}
```

**Comment by davidlt** [comment]
`std::string` has `operator=(string_view)` and `operator=(const char *)`.

http://en.cppreference.com/w/cpp/string/basic_string/operator%3D

**Comment by pcanal** [comment]
> std::string has operator=(string_view) and operator=(const char *).

Yes?  I.e. also has operator=(string) and I suspect in this case this the one that is picked.  Either way gcc7 with the master and the code I shown does *not* complain about any ambiguity.

**Comment by pcanal** [comment]
> I just checked issues with optional and they are caused because rootcling_stage1 runs in C++14 mode, but no C++17. Currently I don't see where you set a language options for rootcling_stage1.

Apriori it is the same place inside cling .... Are you sure it is not simply that our clang version is too old?

**Comment by davidlt** [comment]
Nope, because my change enable C++1z in Cling (otherwise CMSSW wouldn't work). If I print __cplusplus from rootcling_stage1 I get C++14, which would explain the error about optional, I guess.

**Comment by pcanal** [comment]
>  If I print __cplusplus from rootcling_stage1
How did you print it?

**Comment by Axel-Naumann** [comment]
Indeed, we only have (in interpreter/cling/lib/Interpreter/CIFactory.cpp):
```
#if __cplusplus > 201103L
      Opts.CPlusPlus14 = 1;
#endif
#if __cplusplus >= 201103L
      Opts.CPlusPlus11 = 1;
#endif
```
Our clang has `CPlusPlus1z`; maybe add that to the patch?

Axel.

**Comment by pcanal** [comment]
it is added by the patch:
```
#if __cplusplus >= 201703L
      Opts.CPlusPlus1z = 1;
#endif
```

**Comment by pcanal** [comment]
my test also shows that the line is executed but it is ineffective because of:
```
    // FIXME: Use correct value for C++17.
    if (LangOpts.CPlusPlus1z)
      Builder.defineMacro("__cplusplus", "201406L");
```
llvm/src/tools/clang/lib/Frontend/InitPreprocessor.cpp

**Comment by davidlt** [comment]
```
diff --git a/core/clingutils/src/unordered_setLinkdef.h b/core/clingutils/src/unordered_setLinkdef.h
index a9dc6eb..fd26107 100644
--- a/core/clingutils/src/unordered_setLinkdef.h
+++ b/core/clingutils/src/unordered_setLinkdef.h
@@ -1,6 +1,11 @@
 #include <unordered_set>
 #include <string>

+#define XSTR(x) STR(x)
+#define STR(x) #x
+
+#pragma message "Version " XSTR(__cplusplus)
+
 #pragma create TClass unordered_set<int>;
 #pragma create TClass unordered_set<long>;
 #pragma create TClass unordered_set<float>;
```

I get `201406L`.

Huh from CMSSW, GCC7, ROOT:
```
root [0] __cplusplus
(long) 201406
```

I guess, https://github.com/root-project/root/commit/702298d9ad83866d0be62f0422c03ac8ea6687f1 needs improvements.

Clang 3.9 didn't have a correct value for C++17, but __it is in C++17 mode__:
```
 379   } else {
 380     // FIXME: Use correct value for C++17.
 381     if (LangOpts.CPlusPlus1z)
 382       Builder.defineMacro("__cplusplus", "201406L");
 383     // C++1y [cpp.predefined]p1:
 384     //   The name __cplusplus is defined to the value 201402L when compiling a
 385     //   C++ translation unit.
 386     else if (LangOpts.CPlusPlus14)
 387       Builder.defineMacro("__cplusplus", "201402L");
```

Even 4.0.0 doesn't have a correct value. Only master branch of Clang has the correct value.

```
commit 65ecf3afb2faf3d0598a6c742b81e7adec6b101b
Author: Richard Smith <richard-llvm@metafoo.co.uk>
Date:   Mon Mar 20 20:12:48 2017 +0000

    Bump __cplusplus for C++17 to 201703L per the C++17 DIS.
```

Maybe we should adjust __cplusplus value to match official one and the one in GCC.

None of this helps with optional error we have.

**Comment by Axel-Naumann** [comment]
Fair enough. Summary: no C++17 in ROOT 6.10. It will only arrive though a llvm/clang upgrade.

**Comment by davidlt** [comment]
Clang 3.9 has majority of C++17 language features.

I quickly took content of `multimapLinkdef.h` and tested with Clang 3.8.0 and Clang 4.0.0 with GCC 7.1.1. Clang 4.0.0 worked just fine, but Clang 3.8.0 didn't. Sadly I don't have 3.9.0.

**Pull request closed without merging**